Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    // Copy the index since _Py_Identifier.index is declared as volatile
+    Py_ssize_t index = id->index;
+    if (index < 0) {
+        struct _Py_unicode_runtime_ids *rt_ids = &interp->runtime->unicode_ids;
+
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
+        // Check again to detect concurrent access. Another thread can have
+        // initialized the index while this thread waited for the lock.
+        //
+        // _Py_Identifier.index is declared with volatile keyword to ensure
+        // that the index is read again here (prevent compiler optimization).
+        // Use volatile instead of an atomic variable since pycore_atomic.h is
+        // an internal header file, and stdatomic.h is not yet widely adopted.
+        if (id->index < 0) {
+            id->index = rt_ids->next_index;
+            rt_ids->next_index++;
+        }
+        PyThread_release_lock(rt_ids->lock);
+        index = id->index;
+    }
+    assert(index >= 0);
+
+    PyObject *object;
+    if (index < ids->size) {
```

Comment:
I afraid that `ids->size` should be volatile and atomic too.

Reply:
I chose to avoid atomic variables since pycore_atomic.h is an internal header and I don't think that all C compilers supported by Python implement <stdatomic.h>.

Why does it have to be atomic?

Reply:
If you expect a risk that Py_ssize_t cannot be updated atomatically on all architectures, maybe we can replace it with an int. We have around 500 identifiers currently: it fits into INT_MAX :-)

Reply:
It is not guaranteed that `int` or any other type is updated atomically unless you use atomic integers. It is for what they are introduced.

Reply:
`ids->size` is protected by the GIL.

I'm talking about id->index which is volatile but not atomic.

_Py_Identifier.index (id->index) is protected by `_PyRuntime.unicode_ids.lock` which IMO is enough to avoid having to declare _Py_Identifier.index as atomic.